src/MergeContext.js

```diff
@@ -499,58 +559,115 @@ class MergeContext {
+            return null;
+        }
+
+        return requiredContexts;
+    }
+
+    // returns filled StatusResult object
+    async _getPRStatuses() {
+        const requiredContexts = await this._getRequiredContexts();
+        if (!requiredContexts)
+            return new StatusResult(0, "PR");
+
+        let combinedStatus = await GH.getStatuses(this._prHeadSha());
+
+        let statusResult = new StatusResult(requiredContexts.length, "PR");
         // filter out non-required checks
         for (let st of combinedStatus.statuses) {
-            if (requiredContexts.find((el) => { return needMatching ?
-                         st.context.startsWith(el.trim()) : (el.trim() === st.context.trim());}))
-                requiredChecks.push({context: st.context, state: st.state, targetUrl: st.target_url, description: st.description});
-        }
-
-        if (requiredChecks.length < requiredChecksNumber || requiredChecks.find(check => check.state === 'pending'))
-            return 'pending';
-
-        const prevLen = requiredChecks.length;
-        requiredChecks = requiredChecks.filter(check => check.state === 'success');
-        const ret = (prevLen === requiredChecks.length);
-        if (ret && andSupplyRequired) {
-            for (let requiredContext of requiredContexts) {
-                // go further only if the passed check context is not a required one
-                if (requiredChecks.find(el => el.context.trim() === requiredContext.trim()) === undefined) {
-                    // go further only if the passed check context matches the required one
-                    const matched = requiredChecks.find(el => el.context.startsWith(requiredContext.trim()));
-                    // Before a 'staged' commit can be applied, it should have all required checks passed.
-                    // We create a new "required" check, taking other attributes like targetUrl
-                    // from an already succeeded (matching) check (there can be several matching checks,
-                    // e.g., from different Jenkins nodes). After that, there will be two checks
-                    // referencing the same targetUrl.
-                    if (matched && !this._dryRun("required status check creation"))
-                        await GH.createStatus(ref, "success", matched.targetUrl, matched.description, requiredContext);
-                }
+            if (requiredContexts.find(el => el.trim() === st.context.trim()) !== undefined)
+                statusResult.addRequiredStatus(new StatusCheck(st.context, st.state, st.target_url, st.description));
+        }
+        return statusResult;
+    }
+
+    async _checkPRStatuses(logContext) {
+        const commitStatus = await this._getPRStatuses();
+        this._log("status details: " + commitStatus);
+        if (commitStatus.succeeded())
+            return StepResult.Succeed();
+        else if (this._approval.grantedTimeout() && commitStatus.approvalPending())
```

Comment:
We do not wait if a test has failed: commitStatus.approvalPending() should return false in this case.

Reply:
Why should approvalPending() return false when an approval is pending? Approvals have nothing to do with failed tests.

Reply:
> Eduard: commitStatus.approvalPending() should return false if a test has failed
> Alex: Why should approvalPending() return false when an approval is pending?
> Eduard: This statement is not correct

My statement is implied by yours. I am just asking why we should use such odd logic.

> approvalPending() returns true if and only if all passed tests succeeded except "PR approval" test which is pending.

Yes. The first condition looks unrelated to the method (triggering my question), and the second condition should probably be based on our approval calculation rather than GitHub status check (but that is a separate problem).

This code is gone in the latest revision, right? If so, we do not need to discuss it anymore.